src/dmd/canthrow.d

```diff
@@ -82,6 +82,19 @@ extern (C++) bool canThrow(Expression e, FuncDeclaration func, bool mustNotThrow
 
             if (global.errors && !ce.e1.type)
                 return; // error recovery
+
+            import dmd.id : Id;
+
+            if (ce.f && ce.f.ident == Id._d_delstruct)
+            {
+                // In expressionsem.d `delete s` was lowered to `_d_delstruct(s)`.
+                // The following code rewrites it back to the original expresion
+                // in order to properly output `nothrow` errors.
+                auto de = new DeleteExp(ce.loc, (*ce.arguments)[0], false);
```

Comment:
Creating a DeleteExp is superfluous here. This is wasteful because allocation is costly. You can simply check if the dtor of of the aggregate type (if it exists) throws: this is essentially what the DeleteExp visiting method does.

Reply:
I was trying to not duplicate the code in `visit(DeleteExp)`, but I see your point.